src/service-worker.ts

```diff
@@ -48,4 +58,33 @@ registerRoute(({ request, url }: { request: Request; url: URL }) => {
   }
 
   return true
-}, createHandlerBoundToURL(process.env.PUBLIC_URL + '/index.html'))
+}
+
+const APP_SHELL_HTML = process.env.PUBLIC_URL + '/index.html'
+let cachedAppShell: Response | undefined
+
+async function handleAppShell({ request, event }: RouteHandlerCallbackOptions) {
+  // NB: If the cached response is used, the preloadResponse may be cancelled before it is settled.
+  // This will log an error to the console, but it can be ignored - it *should* be cancelled.
+  const { preloadResponse } = event as unknown as { preloadResponse: Promise<Response | undefined> }
+  const controller = new AbortController()
+  const response = (await preloadResponse) || (await fetch(APP_SHELL_HTML, { signal: controller.signal }))
+
+  // Only serve from the cache if the cache is still fresh (checked using etag comparison). The etag
+  // header can be queried before the response has streamed, so this still reduces network time.
+  // (In manual testing, etag lookup took only 2-20% of the total response download time.)
+  const etag = response.headers.get('etag')
+  const cachedEtag = cachedAppShell?.headers.get('etag')
+  if (cachedAppShell && (isLocal || etag) && etag === cachedEtag) {
+    // If the cache is still fresh, cancel any pending response. The preloadResponse is cancelled
+    // automatically by returning before it is settled.
+    controller.abort()
+  } else {
+    cachedAppShell = response
+  }
+  // TODO(leggechr): Send beacons with metrics to track usage/load times. This should probably be done on the client, not the serviceworker.
+  // Cached responses can have a header added to indicate that they were served from cache.
+
+  const clone = cachedAppShell.clone()
```

Comment:
How come we have to clone here?

Reply:
Responses are one-off - once they are read they are no longer valid. Clone allows copying them without invalidating them, so it's necessary for a cache, where the same responses may be read many times. Without it, you'd get a TypeError. 

See https://developer.mozilla.org/en-US/docs/Web/API/Request/clone (particularly the second paragraph).